### ISSUE TITLE
[11.0] Error when input fields One2many (Tree editable)

### ISSUE THREAD

**Issue by dntq83** [odoo/odoo#24936]
Impacted versions:
11.0
Steps to reproduce:
When input data, an error occurs if press tab key in a field next to a read-only field (tree editable in form)
Current behavior:
Error accurs
Expected behavior:

Video/Screenshot link (optional):
https://youtu.be/3sSXBYNXd5U

**Comment by pedrobaeza** [comment]
Can you reproduce the problem in runbot server? Do you have any extra module apart from Odoo core?

**Comment by dntq83** [comment]
I can't reproduce in runbot server.
I have tried many times in Odoo Server (odoo core) without adding any modules.

**Comment by dntq83** [comment]
@pedrobaeza  please check code below
`http://10.10.0.1:8119/web/static/src/js/core/dom.js:167
Traceback:
TypeError: Cannot read property 'selectionStart' of undefined
    at Object.getSelectionRange (http://10.10.0.1:8119/web/static/src/js/core/dom.js:167:25)
    at Class.confirmUpdate (http://10.10.0.1:8119/web/static/src/js/views/list/list_editable_renderer.js:146:38)
    at Class.reset (http://10.10.0.1:8119/web/static/src/js/fields/relational_fields.js:790:38)
    at Class.prototype.(anonymous function) [as reset] (http://10.10.0.1:8119/web/static/src/js/core/class.js:90:38)
    at Class.reset (http://10.10.0.1:8119/web/static/src/js/fields/relational_fields.js:1185:28)
    at Class.prototype.(anonymous function) [as reset] (http://10.10.0.1:8119/web/static/src/js/core/class.js:90:38)
    at http://10.10.0.1:8119/web/static/src/js/views/basic/basic_renderer.js:110:34
    at Function._.each._.forEach (http://10.10.0.1:8119/web/static/lib/underscore/underscore.js:145:9)
    at Class.confirmChange (http://10.10.0.1:8119/web/static/src/js/views/basic/basic_renderer.js:107:11)
    at Class.confirmChange (http://10.10.0.1:8119/web/static/src/js/views/form/form_renderer.js:108:28)`

**Comment by pedrobaeza** [comment]
@ged-odoo do you think this is a legitimate issue on Odoo code?

**Comment by ged-odoo** [comment]
the weird thing is that @dntq83 could not reproduce it on the runbot.  However, judging from the stacktrace and the code, it looks like the web client tried to work with the focused element of a readonly widget, which is undefined.  This really look like a web client issue.

**Comment by dntq83** [comment]
@ged-odoo Could you please show me how to simulate input with tab key in runbot? Thank you

**Comment by ljsalvatierra-binovo** [comment]
[v11.0]
I get a similar error when i hit the tab key and the next field is a readonly monetary or a float field.
For what i see in the code:
https://github.com/odoo/odoo/blob/11.0/addons/web/static/src/js/fields/basic_fields.js#L191
```js
    getFocusableElement: function () {
        return this.$input || $();
    }
```
Because of MonetaryField does not override the function like others, it returns undefined.
Same happens with FieldFloat.
```
http://ljsalvatierra-desktop:11069/web/static/src/js/core/dom.js:167
Traceback:
getSelectionRange@http://ljsalvatierra-desktop:11069/web/static/src/js/core/dom.js:167:13
confirmUpdate@http://ljsalvatierra-desktop:11069/web/static/src/js/views/list/list_editable_renderer.js:146:38
reset@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/relational_fields.js:798:38
OdooClass.extend/</prototype[name]</<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/class.js:90:38
reset@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/relational_fields.js:1193:28
OdooClass.extend/</prototype[name]</<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/class.js:90:38
confirmChange/<@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_renderer.js:110:34
_.forEach@http://ljsalvatierra-desktop:11069/web/static/lib/underscore/underscore.js:145:9
confirmChange@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_renderer.js:107:11
confirmChange@http://ljsalvatierra-desktop:11069/web/static/src/js/views/form/form_renderer.js:108:28
OdooClass.extend/</prototype[name]</<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/class.js:90:38
confirmChange@http://ljsalvatierra-desktop:11069/mail/static/src/js/form_renderer.js:39:28
OdooClass.extend/Class.include/</prototype[name]</<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/class.js:123:38
_confirmChange@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_controller.js:282:30
_applyChanges/<@http://ljsalvatierra-desktop:11069/web/static/src/js/views/field_manager_mixin.js:57:33
then/</</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3276:89
fire@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3119:58
add@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3165:49
then/</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3275:77
each@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:383:58
then/<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3272:56
Deferred@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3332:30
then@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3271:55
_applyChanges@http://ljsalvatierra-desktop:11069/web/static/src/js/views/field_manager_mixin.js:51:14
_applyChanges/<@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_controller.js:217:20
exec/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/concurrency.js:216:31
then/</</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3276:89
fire@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3119:58
add@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3165:49
then/</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3275:77
each@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:383:58
then/<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3272:56
Deferred@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3332:30
then@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3271:55
exec@http://ljsalvatierra-desktop:11069/web/static/src/js/core/concurrency.js:215:28
_applyChanges@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_controller.js:216:27
OdooClass.extend/</prototype[name]</<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/class.js:90:38
_onFieldChanged@http://ljsalvatierra-desktop:11069/web/static/src/js/views/field_manager_mixin.js:106:14
_onFieldChanged@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_controller.js:538:43
proxy/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:279:23
trigger@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:226:31
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:339:34
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:341:20
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:341:20
trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:335:14
_setValue@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/abstract_field.js:430:14
_onFieldChanged@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/relational_fields.js:1066:18
proxy/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:279:23
trigger@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:226:31
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:339:34
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:341:20
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:341:20
trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:335:14
_setValue@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/abstract_field.js:430:14
_doAction@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/basic_fields.js:139:18
_onChange@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/basic_fields.js:282:14
proxy/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:279:23
dispatch@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4641:58
add/elemData.handle@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4309:63
trigger@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4816:54
trigger@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4516:74
trigger/<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:5260:38
each@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:383:58
each@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:136:31
trigger@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:5259:29
jQuery.fn[name]@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:8443:30
.focus</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery.ui/jquery-ui.js:230:10
activate@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/abstract_field.js:219:24
_activateFieldWidget@http://ljsalvatierra-desktop:11069/web/static/src/js/views/basic/basic_renderer.js:187:57
_selectCell/<@http://ljsalvatierra-desktop:11069/web/static/src/js/views/list/list_editable_renderer.js:641:31
then/</</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3276:89
fire@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3119:58
add@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3165:49
then/</<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3275:77
each@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:383:58
then/<@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3272:56
Deferred@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3332:30
then@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:3271:55
_selectCell@http://ljsalvatierra-desktop:11069/web/static/src/js/views/list/list_editable_renderer.js:631:42
_onNavigationMove@http://ljsalvatierra-desktop:11069/web/static/src/js/views/list/list_editable_renderer.js:794:26
proxy/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:279:23
trigger@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:226:31
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:339:34
_trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:341:20
trigger_up@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:335:14
_onKeydown@http://ljsalvatierra-desktop:11069/web/static/src/js/fields/abstract_field.js:464:22
proxy/<@http://ljsalvatierra-desktop:11069/web/static/src/js/core/mixins.js:279:23
dispatch@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4641:58
add/elemData.handle@http://ljsalvatierra-desktop:11069/web/static/lib/jquery/jquery.js:4309:63
```

**Comment by ljsalvatierra-binovo** [comment]
I've done some more digging and found this:
On commit 36b4d78a2678806b25aaa5650b64496f9da9d205 addons/web/static/src/js/views/list/**list_editable_renderer.js** on function **_confirmUpdate_**
The **focusedElement** is obtained whether **isFocusable** or not.

```js
        // store the cursor position to restore it once potential onchanges have
        // been applied
        var currentRowID, currentWidget, focusedElement, selectionRange;
        if (self.currentRow !== null) {
            currentRowID = this.state.data[this.currentRow].id;
            currentWidget = this.allFieldWidgets[currentRowID][this.currentFieldIndex];
            focusedElement = currentWidget.getFocusableElement().get(0);
            if (currentWidget.formatType !== 'boolean') {
                selectionRange = dom.getSelectionRange(focusedElement);
            }
        }
...
                return self._selectCell(newRowIndex, self.currentFieldIndex, {force: true}).then(function () {
                    // restore the cursor position
                    currentRowID = self.state.data[newRowIndex].id;
                    currentWidget = self.allFieldWidgets[currentRowID][self.currentFieldIndex];
                    focusedElement = currentWidget.getFocusableElement().get(0);
                    if (selectionRange) {
                        dom.setSelectionRange(focusedElement, selectionRange);
                    }
                });
            }
        });
    },
```

Also, for the testing i've done, even the selectionRange is not correctly calculated.
Example:

First field | Second field
------------ | ---------------
Some text | 0.00€

If i change the second field when the first field changes:
* (Char) Some text -> Some other text
* (Monetary) 0.00€ -> 100,000.00€

First field | Second field
------------ | ---------------
Some other text | 100,000.00€

It selects only the previous range, three characters in this case.
![image](https://user-images.githubusercontent.com/23120306/60664168-7d236a00-9e61-11e9-901b-d9fa40c72de9.png)
![image](https://user-images.githubusercontent.com/23120306/60664119-5b29e780-9e61-11e9-93ae-36bcc11383cb.png)

**Comment by ljsalvatierra-binovo** [comment]
I've created a pull request on odoo/odoo, should i make here too? https://github.com/odoo/odoo/pull/34607

**Comment by Yenthe666** [comment]
@ljsalvatierra-binovo PR's are always welcomed and have a higher priority than issues.

**Comment by ibtissem** [comment]
I'm facing the same problem, is this issue fixed ? I'm working on V11.0 , I'm facing the exception when tabbing key and the next field is a computed readonly float field. thanks

**Comment by C3POdoo** [comment]
Dear @dntq83,

Thank you for your report but we are closing it due to inactivity.
We apology if we could not look at your request in time.
If your report still makes sense, don't hesitate to reopen a new one. We will try to check it as soon as possible.

_This is an automated message._
